src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -395,7 +395,10 @@ impl<O: ForestObligation> ObligationForest<O> {
         let mut errors = vec![];
         let mut stalled = true;
 
-        for index in 0..self.nodes.len() {
+        // Note that the loop body can append new nodes, and those new nodes
+        // will then be processed by subsequent iterations of the loop.
+        let mut index = 0;
+        while index < self.nodes.len() {
```

Comment:
I don't see any improvements here. In fact it's a step backwards and it should be `for node in self.nodes.iter_mut()`?

Reply:
The commit has this comment to the top of the loop:
```
        // Note that the loop body can append new nodes, and those new nodes
        // will then be processed by subsequent iterations of the loop.
```
(The PR description also mentions this.) As a result, an iterator cannot be used.

Reply:
I clarified the comment at the top of the loop.